update WebApi to .NET 9 (without Swagger) (#34746)
dotnet/AspNetCore.Docs: aspnetcore/tutorials/first-web-api.md

[!INCLUDE[](~/includes/package-reference.md)]

### Run the project

The project template creates a `WeatherForecast` API with support for [OpenAPI](xref:fundamentals/openapi/overview).

# [Visual Studio](#tab/visual-studio)

Press Ctrl+F5 to run without the debugger.

[!INCLUDE[](~/includes/trustCertVS.md)]

> Visual Studio launches a terminal window and prints out the URL of the running application. The API is hosted at `https://localhost:<port>`, where `<port>` is a randomly chosen port number set at the project creation. 

Reviewer: ```suggestion
Visual Studio launches a terminal window and displays the URL of the running app. The API is hosted at `https://localhost:<port>`, where `<port>` is a randomly chosen port number set at the project creation. 
```

* <kbd>Ctrl</kbd>+*click* the HTTPS URL in the output to test the web app in a browser.

* The default browser is launched to `https://localhost:<port>`, where `<port>` is the randomly chosen port number displayed in the output. There's no endpoint at `https://localhost:<port>`, so the browser returns [HTTP 404 Not Found](https://developer.mozilla.org/docs/Web/HTTP/Status/404).

* Append `/weatherforecast` to the URL to test the WeatherForecast API. The browser displays JSON similar to the following example:
    
```json
[
    {
        "date": "2025-07-16",
        "temperatureC": 52,
        "temperatureF": 125,
        "summary": "Mild"
    },
    {
        "date": "2025-07-17",
        "temperatureC": 36,
        "temperatureF": 96,
        "summary": "Warm"
    },
    {
        "date": "2025-07-18",
        "temperatureC": 39,
        "temperatureF": 102,
        "summary": "Cool"
    },
    {
        "date": "2025-07-19",
        "temperatureC": 10,
        "temperatureF": 49,
        "summary": "Bracing"
    },
    {
        "date": "2025-07-20",
        "temperatureC": -1,
        "temperatureF": 31,
        "summary": "Chilly"
    }
]
```
    

> * After testing the web app in the following instruction, press <kbd>Ctrl</kbd>+<kbd>C</kbd> in the integrated terminal to shut it down.

Reviewer: ```suggestion
* After testing the web app using the following instruction, press <kbd>Ctrl</kbd>+<kbd>C</kbd> in the integrated terminal to close it.
```

---

### Test the project

> # [Visual Studio](#tab/visual-studio)

Reviewer: ```suggestion
<!--
# [Visual Studio](#tab/visual-studio) -->
```

This tutorial uses [Endpoints Explorer and .http files](xref:test/http-files#use-endpoints-explorer) to test the API.

> 

Reviewer 2: Do you mean "Select the Visual Studio **Code"** tab?  They are already on the VS tab at this point.  I think it would make things a little confusing.  They could get through it, but it feels a bit awkward.

> 

Reviewer: > Do you mean to write "Select the Visual Studio **Code"** tab? They are already on the VS tab at this point. I think it would make things a little confusing. They could get through it, but it feels a bit awkward.

Yes, forgot ***Code***,
so there's no value in VSC customers doing that section?

> 

Reviewer 2: There's value, minus a bit of awkward.  I ran through at that point from VS to VS Code instruction, it works.  Consider making it clear that the Swagger option is instead.  Also cap "Swagger"

> 

Author: To be honest, I also find the current wording/format unclear.
If we change this, we should also re-evaluate the minimal api docs (https://learn.microsoft.com/en-us/aspnet/core/tutorials/min-web-api?view=aspnetcore-9.0&tabs=visual-studio-code#add-the-api-code), which uses the same format.

I believe the reasoning behind the current format is that Visual Studio supports http files by default, and thus swagger isn't required.
For Visual Studio Code this is not the case, an extension needs to be installed to use http files. That's why I assume the Swagger setup is added to the Visual Studio Code tabs.

I like the suggestion from @Rick-Anderson .
An alternative is to remove the Visual Studio/Visual Studio Code setup and suggest users to install the http files extension, or point them to the "Test with other tools" section for alternatives.

> # [Visual Studio Code](#tab/visual-studio-code)

Reviewer: ```suggestion
<!--
# [Visual Studio Code](#tab/visual-studio-code) -->
```

## Create API testing UI with Swagger

> There are many available web API testing tools to choose from, and you can follow this tutorial's introductory API test steps with your own preferred tool.

Reviewer: ```suggestion
There are many available web API testing tools to choose from, and you can follow this tutorial's introductory API test steps with your preferred tool.
```

This tutorial utilizes the .NET package [NSwag.AspNetCore](https://www.nuget.org/packages/NSwag.AspNetCore/), which integrates Swagger tools for generating a testing UI adhering to the OpenAPI specification:

* NSwag: A .NET library that integrates Swagger directly into ASP.NET Core applications, providing middleware and configuration.
* Swagger: A set of open-source tools such as OpenAPIGenerator and SwaggerUI that generate API testing pages that follow the OpenAPI specification.
* OpenAPI specification: A document that describes the capabilities of the API, based on the XML and attribute annotations within the controllers and models.

For more information on using OpenAPI and NSwag with ASP.NET, see <xref:tutorials/web-api-help-pages-using-swagger>.

### Install Swagger tooling

* Run the following command:

  ```dotnetcli
  dotnet add package NSwag.AspNetCore

>   ```

Reviewer: I think this is where you need `# [Visual Studio Code](#tab/visual-studio-code) ` and `VSC`

Reviewer 2: @Rick-Anderson, NSwag/Swagger setup is only being used for VSC here.   The VSC tab is set currently for the entire section here: ## Create API testing UI with Swagger

That seems right the way the tabs are set here. Maybe I don't understand what you are referring to.

The previous command adds the [NSwag.AspNetCore](https://www.nuget.org/packages/NSwag.AspNetCore/) package, which contains tools to generate Swagger documents and UI.
Because our project is using OpenAPI, we only use the NSwag package to generate the Swagger UI.

### Configure Swagger middleware

> * In Program.cs add the following highlighted code after the `app` is defined

Reviewer: ```suggestion
* In `Program.cs`, add the following highlighted code:
```

:::code language="csharp" source="~/tutorials/first-web-api/samples/9.0/TodoApi_SwaggerVersion/Program.cs" id="snippet_UseSwagger" highlight="6-9":::

The previous code enables the Swagger middleware for serving the generated JSON document using the Swagger UI. Swagger is only enabled in a development environment. Enabling Swagger in a production environment could expose potentially sensitive details about the API's structure and implementation.

> To generate the UI, it uses the OpenAPI document generated by OpenApi, located at `/openapi/v1.json`.

Reviewer: ```suggestion
The app uses the OpenAPI document generated by OpenApi, located at `/openapi/v1.json`, to generate the UI.
```

> View the generated OpenAPI specification for the `WeatherForecast` API while the project is running by navigating your browser to `https://localhost:<port>/openapi/v1.json`.

Reviewer: ```suggestion
View the generated OpenAPI specification for the `WeatherForecast` API while the project is running by navigating  to `https://localhost:<port>/openapi/v1.json` in your browser.
```

* Append `/weatherforecast` to the URL to test the WeatherForecast API. The browser displays JSON similar to the following example:
    
```json
[
    {
        "date": "2025-07-16",
        "temperatureC": 52,
        "temperatureF": 125,
        "summary": "Mild"
    },
    {
        "date": "2025-07-17",
        "temperatureC": 36,
        "temperatureF": 96,
        "summary": "Warm"
    },
    {
        "date": "2025-07-18",
        "temperatureC": 39,
        "temperatureF": 102,
        "summary": "Cool"
    },
    {
        "date": "2025-07-19",
        "temperatureC": 10,
        "temperatureF": 49,
        "summary": "Bracing"
    },
    {
        "date": "2025-07-20",
        "temperatureC": -1,
        "temperatureF": 31,
        "summary": "Chilly"
    }
]
```
    
* After testing the web app using the following instruction, press <kbd>Ctrl</kbd>+<kbd>C</kbd> in the integrated terminal to close it.

---

### Test the project

> <!--

Reviewer: ```suggestion
```

# [Visual Studio](#tab/visual-studio) -->

This tutorial uses [Endpoints Explorer and .http files](xref:test/http-files#use-endpoints-explorer) to test the API.

> <!--
> # [Visual Studio Code](#tab/visual-studio-code) -->

Reviewer: ```suggestion
# [Visual Studio Code](#tab/visual-studio-code) 
```

### Configure Swagger middleware

* In `Program.cs`, add the following highlighted code:

:::code language="csharp" source="~/tutorials/first-web-api/samples/9.0/TodoApi_SwaggerVersion/Program.cs" id="snippet_UseSwagger" highlight="6-9":::

The previous code enables the Swagger middleware for serving the generated JSON document using the Swagger UI. Swagger is only enabled in a development environment. Enabling Swagger in a production environment could expose potentially sensitive details about the API's structure and implementation.

The app uses the OpenAPI document generated by OpenApi, located at `/openapi/v1.json`, to generate the UI.
View the generated OpenAPI specification for the `WeatherForecast` API while the project is running by navigating to `https://localhost:<port>/openapi/v1.json` in your browser.

The OpenAPI specification is a document in JSON format that describes the structure and capabilities of your API, including endpoints, request/response formats, parameters, and more. It's essentially a blueprint of your API that can be used by various tools to understand and interact with your API.

> <!--
> ---
> -->

Reviewer: ```suggestion
---
```

  * The triple hashtag (`###`) line is a request delimiter: what comes after it is for a different request.

* The POST request expects a `TodoItem`. To define the todo, replace the `//TodoItem` comment with the following JSON:

  ```json
  {
    "name": "walk dog",
    "isComplete": true
  }
  ```
  
  The TodoApi.http file should now look like the following example, but with your port number:
  
  ```
  @TodoApi_HostAddress = https://localhost:7260
  
  Post {{TodoApi_HostAddress}}/api/todoitems
  Content-Type: application/json
  
  {
    "name": "walk dog",
    "isComplete": true
  }
  
  ###
  ```

* Run the app.

* Select the **Send request** link that is above the `POST` request line.

  ![.http file window with run link highlighted.](~/tutorials/first-web-api/_static/9/http-file-run-button-vs17.13.0.png)

  The POST request is sent to the app and the response is displayed in the **Response** pane.

  ![.http file window with response from the POST request.](~/tutorials/first-web-api/_static/9/http-file-window-with-response-vs17.13.0.png)

# [Visual Studio Code](#tab/visual-studio-code)

> * With the app still running, in the browser, navigate to `https://localhost:<port>/swagger` to display the API testing page generated by Swagger. Click on "TodoItems" to expand the operations.

Reviewer: ```suggestion
* With the app still running, in the browser, navigate to `https://localhost:<port>/swagger` to display the API testing page generated by Swagger. Click on **TodoItems** to expand the operations.
```

To demonstrate the DTO approach, update the `TodoItem` class to include a secret field:

:::code language="csharp" source="~/tutorials/first-web-api/samples/9.0/TodoApiDTO/Models/TodoItem.cs" highlight="8" :::

The secret field needs to be hidden from this app, but an administrative app could choose to expose it.

Verify you can post and get the secret field.

> Create a DTO model:

Reviewer 2: ```suggestion
Create a DTO model in a **Models/TodoItemsDTO.cs** file:
```
